pkg/gitops/templates.go

```diff
@@ -0,0 +1,69 @@
+package gitops
+
+import (
+	"fmt"
+	"html/template"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+)
+
+//go:generate moq -out templates_moq_test.go . renderAllFileser
+type renderAllFileser interface {
+	renderAllFiles() error
+}
+
+// templatesRenderer implements the renderAllFileser interface.
+var _ sshKeyer = (*sshKey)(nil)
+
+// TemplatesRenderer renders a folder of templates to a local repository.
+type TemplatesRenderer struct {
+	// Source folder of templates.
+	SourceFolder string
+	// Variables to substitute into the templates.
+	Vars map[string]string
+	// Destination repository for rendered files.
+	DestinationRepo repositorier
+	// Destination folder inside the repository for rendered files.
+	DestinationFolder string
+}
+
+func (tr TemplatesRenderer) renderAllFiles() error {
+	// Get all template file names from the source folder.
+	files, err := ioutil.ReadDir(tr.SourceFolder)
+	if err != nil {
+		return fmt.Errorf("read files in %q: %w", tr.SourceFolder, err)
+	}
+
+	// Render templates one-by-one to the destinaton folder
+	// (substituting variables given).
+	for _, file := range files {
+		if err := tr.renderFile(file.Name()); err != nil {
+			return fmt.Errorf("render file %q: %w", file.Name(), err)
+		}
+	}
+	return nil
+}
+
+func (tr TemplatesRenderer) renderFile(fileName string) error {
+	// Parse template.
+	sourceFilePath := filepath.Join(tr.SourceFolder, fileName)
+	t, err := template.ParseFiles(sourceFilePath)
+	if err != nil {
+		return fmt.Errorf("parse template %q: %w", sourceFilePath, err)
+	}
+
+	// Create a file for the rendered template.
+	destinationFilePath := filepath.Join(
+		tr.DestinationRepo.localPath(), tr.DestinationFolder, fileName)
```

Comment:
Why do we need the entire `DestinationRepo` if we only use the path?

Reply:
Previously only the local repo path was passed into the renderer. I found this a bit better because that way `main.go` had to create the local repo clone and retrieve it's temporary path passing it into the renderer.

Now `main.go` is like: here is the temporary repository, render the files into it please (`main.go` has less business logic inside it, it's pretty much glue code only).

Reply:
I meant something like this:
```
type TemplatesRender struct {
         DestinationPath string
	// Destination folder inside the repository for rendered files.
	DestinationFolder string
...
}

in main():
        // Create templates renderer.
	renderer := gitops.TemplatesRenderer{
		SourceFolder:      cfg.TemplatesFolder,
		Vars:              cfg.Vars,
		DestinationPath:   repo.localPath(),
		DestinationFolder: cfg.DeployFolder,
	}
```

Reply:
Yes, this is the exact way it was before. Let's discuss this in person next week as well and update this thread with the conclusion :)